code/parachain/frame/pablo/src/dual_asset_constant_product.rs

```diff
@@ -60,44 +64,54 @@ impl<T: Config> Uniswap<T> {
 	}
 
 	pub(crate) fn get_exchange_value(
-		pool: &ConstantProductPoolInfo<T::AccountId, T::AssetId>,
+		pool: &BasicPoolInfo<T::AccountId, T::AssetId, ConstU32<2>>,
 		pool_account: &T::AccountId,
-		asset_id: T::AssetId,
-		amount: T::Balance,
+		in_asset_id: T::AssetId,
+		in_asset_amount: T::Balance,
 	) -> Result<T::Balance, DispatchError> {
-		ensure!(pool.pair.contains(asset_id), Error::<T>::InvalidAsset);
-		let amount = T::Convert::convert(amount);
-		let pool_base_aum = T::Convert::convert(T::Assets::balance(pool.pair.base, pool_account));
-		let pool_quote_aum = T::Convert::convert(T::Assets::balance(pool.pair.quote, pool_account));
+		ensure!(pool.assets_weights.contains_key(&in_asset_id), Error::<T>::InvalidAsset);
+		let amount = T::Convert::convert(in_asset_amount);
+		let pool_assets = Self::get_pool_balances(pool, pool_account);
+		// TODO (vim): We have no way of knowing which amount is for which asset (fixed in a later
+		//  stage). For now we assume the pool defined order which is wrong as the order is based on
+		// Ord
+		let base_asset = pool_assets[0];
+		let quote_asset = pool_assets[1];
 		ensure!(
-			!pool_base_aum.is_zero() && !pool_quote_aum.is_zero(),
+			pool_assets.iter().find(|(_, _, balance)| balance.is_zero()).is_none(),
 			Error::<T>::NotEnoughLiquidity
 		);
+
 		// TODO (vim): Following does not work for "buy" as this causes "out_given_in" to be used in
 		//  case user wanting to buy the quote asset of the pool.
-		let exchange_amount = if asset_id == pool.pair.quote {
-			compute_out_given_in(
-				pool.quote_weight,
-				pool.base_weight,
-				pool_quote_aum,
-				pool_base_aum,
-				amount,
-			)
+		let exchange_amount = if in_asset_id == quote_asset.0 {
+			compute_out_given_in(quote_asset.1, base_asset.1, quote_asset.2, base_asset.2, amount)
 		} else {
-			compute_in_given_out(
-				pool.quote_weight,
-				pool.base_weight,
-				pool_quote_aum,
-				pool_base_aum,
-				amount,
-			)
+			compute_in_given_out(quote_asset.1, base_asset.1, quote_asset.2, base_asset.2, amount)
```

Comment:
Time to upgrade to new methods now that they are in main?

Reply:
That comes in a later task